src/com/google/javascript/jscomp/PeepholeReplaceKnownMethods.java

```diff
@@ -956,4 +958,135 @@ private Node tryFoldStringSplit(Node n, Node stringNode, Node arg1) {
     reportChangeToEnclosingScope(parent);
     return arrayOfStrings;
   }
+
+  private Node tryToFoldArrayConcat(Node n) {
+    checkArgument(n.isCall(), n);
+
+    if (!isASTNormalized() || !useTypes) {
+      return n;
+    }
+    if (isSafeToRemoveArrayLiteralFromFrontOfConcat(n)) {
```

Comment:
It would be good inline the logic of `isConcatCall()` here and use it to pull apart `n` into the bits you'll need and early exit if it turns out `n` isn't a concat call.

Then the utility methods that do the work would take the pieces, which you could check the validity of with not-very-bright `checkArgument()` calls etc.
This reduces redundant logic and improves readability.

```java
Node calleeNode = checkNotNull(n.getFirstChild());
if (!calleeNode.isGetProp()) {
  return n; // not an `array.concat()` call
}
Node callTargetNode = calleeNode.getFirstChild();
String methodName = calleeNode.getLastChild().getString();

if (!"concat".equals(methodName) {
  return n; // not an `array.concat()` call
}
// etc.
```

Reply:
@brad4d  I get your point. Removed redundant logic, moved `isConcatCall` on level up, so if `n` is not concat we exit early.
But I'm not sure if we need to expose the logic of  `isConcatCall`,
because this method used in two places.
 1. To check if it actually concat function call. https://github.com/qqingvarqq/closure-compiler/blob/c9716b3185b90cb7599900907ce5d8752d503953/src/com/google/javascript/jscomp/PeepholeReplaceKnownMethods.java#L965
 2. To check if the previous function call in the chainig is concat function call (mean `.concat().concat()`.https://github.com/qqingvarqq/closure-compiler/blob/c9716b3185b90cb7599900907ce5d8752d503953/src/com/google/javascript/jscomp/PeepholeReplaceKnownMethods.java#L1036

So if I expose the logic it brings code duplication, which is not good.
But if you are sure we need it I can expose this logic.

Reply:
We can avoid both code duplication and redundant execution of code if we replace `isConcatFunctionCall` with a method that returns an object of type `ConcatFunctionCall` or `null` if the node it is passed isn't actually a call to `.concat()`.

```java
abstract private static class ConcatFunctionCall {
  private final Node callNode;
  private final Node calleeNode;
  private final Node firstArgumentNode;
  ConcatFunctionCall(Node callNode, Node calleeNode, Node firstArgumentNode) {
    this.callNode = checkNotNull(callNode);
    this.calleeNode = checkNotNull(calleeNode);
    this.firstArgumentNode = checkNotNull(firstArgumentNode);
  }
}

/**
 * If the argument node is a call to `Array.prototype.concat`, then return a `ConcatFunctionCall` object for it,
 * otherwise return `null`.
 */
@Nullable
private ConcatFunctionCall createConcatFunctionCallForNode(Node n) {
  // logic here to look into n and either return new ConcatFunctionCall or null
}
```

Reply:
Thanks for the detailed explanation.
Updated
Hope now it can be sent on internal review 🤞
By the way, should I squash commits?